src/blueapi/service/runner.py

```diff
@@ -30,83 +28,81 @@ class Runner:
+        self.stop()
+        self.start()
+        LOGGER.info("Runner reloaded")
 
     def start(self):
-        if self._subprocess is None and self._use_subprocess:
-            self._subprocess = Pool(initializer=_init_worker, processes=1)
-            self._subprocess.apply(
-                logging.basicConfig, kwds={"level": self._config.logging.level}
+        try:
+            if self._use_subprocess:
+                self._subprocess = Pool(initializer=_init_worker, processes=1)
+            self.run(setup, [self._config])
+            self._state = EnvironmentResponse(initialized=True)
+        except Exception as e:
+            self._state = EnvironmentResponse(
+                initialized=False,
+                error_message=str(e),
             )
-            try:
-                self._subprocess.apply(start_worker, [self._config])
-            except Exception as e:
-                self._error_message = f"Error configuring blueapi: {e}"
-                LOGGER.exception(self._error_message)
-                return
-            self._initialized = True
-        if not self._use_subprocess:
-            try:
-                start_worker(self._config)
-            except Exception as e:
-                self._error_message = f"Error configuring blueapi: {e}"
-                LOGGER.exception(self._error_message)
-                return
-            self._initialized = True
+            LOGGER.exception(e)
 
     def stop(self):
-        if self._subprocess is not None:
-            self._initialized = False
-            self._subprocess.apply(stop_worker)
-            self._subprocess.close()
-            self._subprocess.join()
-            self._error_message = None
-            self._subprocess = None
-        if (not self._use_subprocess) and (self._initialized or self._error_message):
-            self._initialized = False
-            stop_worker()
-            self._error_message = None
-
-    def reload_context(self):
-        """Reload the subprocess to account for any changes in python modules"""
-        self.stop()
-        self.start()
-        LOGGER.info("Context reloaded")
+        try:
+            self.run(teardown)
+            if self._subprocess is not None:
+                self._subprocess.close()
+                self._subprocess.join()
+            self._state = EnvironmentResponse(
+                initialized=False,
+                error_message=self._state.error_message,
+            )
+        except Exception as e:
+            self._state = EnvironmentResponse(
+                initialized=False,
+                error_message=str(e),
+            )
+            LOGGER.exception(e)
 
     def run(self, function: Callable, arguments: Iterable | None = None) -> Any:
         if self._use_subprocess:
             return self._run_in_subprocess(function, arguments)
         else:
-            if arguments is None:
-                arguments = []
+            arguments = arguments or []
```

Comment:
this function should be refactored to `def run (self, function: Callable[[P], T], arguments: P) -> T:`
Where P = ParamSpec("P"), T = TypeVar("T") if possible

Reply:
Possibly in the future.